Implement `Document` iteration optimization in `proximity.py`

joshfeli opened the pull request:
# Implement `Document` iteration optimization in `proximity.py`

## Overview
This PR implements the optimization suggested in PR #26  implemented in `proximity.py`, which was pasted over from the [Gender Analysis Tookit](https://github.com/dhmit/gender_analysis). Instead of iterating through all `Document` objects associated with [what will be] a `Corpus` instance, this fix delays `QuerySet` evaluation until absolutely needed by making use of its lazy evaluation and the `.values_list()` method.

### Note
This PR assumes that this line exists somewhere in the `Corpus` class implementation:
```python
# import statements
class Corpus(models.Model):
    documents = models.ManyToManyField(Document)
    # other attributes and methods
```

## Next steps:
- *Optimizing iteration in other analysis modules.* The `proximity.py` file is just one of the modules in the `analysis` package; any iteration through the `Document`s of a `Corpus` in other modules would have to be modified.

joshfeli committed:
Copied and pasted code from GATK proximity.py

joshfeli committed:
Labeled a starting point for working with proximity.py

joshfeli committed:
Merge branch 'main' into transfer_proximity_analysis

farooqashar committed:
Updated _generate_token_counter helper function

Updated the _generate_token_counter helper function to reflect the changes of the document and gender models

Co-Authored-By: Joshua Feliciano <41080599+joshfeli@users.noreply.github.com>
Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>
Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

farooqashar committed:
Updated reference to the new label of a document model

Co-Authored-By: Joshua Feliciano <41080599+joshfeli@users.noreply.github.com>
Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>
Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

farooqashar committed:
Efficient looping and results modification revisions

Implemented an efficient way of looping through documents and revised the results dictionary for various analysis functions

Co-Authored-By: Joshua Feliciano <41080599+joshfeli@users.noreply.github.com>
Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>
Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

joshfeli committed:
Reformatted spacing in `proximity.py`

joshfeli committed:
Merge branch 'main' into transfer_proximity_analysis

farooqashar committed:
Added fixtures to add initial data to the pronoun series model

Added json files with initial data

Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>

joshfeli committed:
Optimized performance iterating through `Document`s

Refactored iteration through `Document`s of a `Corpus` such that Django only pulls the relevant fields of each instance from the database rather than each complete `Document` object. We optimized this performance only in `proximity.py`.

Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

joshfeli committed:
Revert "Added fixtures to add initial data to the pronoun series model"

This reverts commit 04adffcb974c0c40588dbe07df326310d522f215.

joshfeli committed:
Merge branch 'main' into optimize_document_iteration

joshfeli committed:
Isolated `run_analysis`; added mechanisms for creating `Document` objects

Import `app.analysis` in your Django shell and call `make_documents(num)` to create the `Document` objects, and call `run_analysis()` with a `QuerySet` [of `Document`s] passed in.

joshfeli committed:
Refined docstring of isolated `run_analysis` function in `__init__.py` in the `analysis` module

joshfeli committed:
Added a variable binding for the query

joshfeli committed:
Trying out a third-party memory profile and server-side cursors in SQL

I'm attempting here to optimize `Document` iteration by implementing a server-side cursor using raw SQL. I'm also using the third-party `memory_profiler` Python package to track any differences in memory usage with this approach. So far, my `DECLARE` query raises a `SyntaxError`.

joshfeli committed:
Added comments to highlight error

joshfeli committed:
Updated comments in `__init__.py` of `analysis` module

joshfeli committed:
Merge branch 'main' into optimize_document_iteration

joshfeli committed:
Merge branch 'main' into optimize_document_iteration

joshfeli committed:
Added `psutil` to `requirements.txt`

The `psutil` package isn't used directly, but it is a dependency for the `memory-profiler` package.

joshfeli committed:
Continuing to experiment with memory profiling

(WORK IN PROGRESS!)
I've added the `eliot_middlemarch.txt` file and have used it instead of `aarund_longfrock.txt` since the former is 1.7MB while the latter is only a couple of KB. I'm using the `@profile` decorator alongside printing the length of the `connection.queries` list after each line to check where/when I'm hitting the DB as well  as how much memory each hit costs.

joshfeli committed:
Corrected `make_documents` so each `Document`'s text is tokenized

This is a crucial correction since the `tokenized_text` field in each Model instance was empty until now!

joshfeli committed:
Added and corrected docstrings to functions in `app/analysis/__init__.py`

joshfeli committed:
Modified docstrings further in `app/analysis/__init__.py`

joshfeli committed:
Trying another approach to optimization

Adding more `Document`s, trying a different approach of querying DB (extreme case of making a new SQL query for each `Document`.

This supposedly is not best practice, but I'm trying it out to see if there are any memory spikes.

I'm also playing with the `gc` built-in module to track [garbage collection](https://docs.python.org/3/library/gc.html).

joshfeli committed:
Mimicking a situation that will require `Document` iteration

Iterating through `Document`s in a "global" frame (i..e in the `analysis_wrapper` function), using each `Document` instance for analysis and calling an analysis function for each instance, returning the results in the "global" frame.

joshfeli committed:
Added some comments to help explain the optimization

joshfeli committed:
Implemented `Document` iteration optimization in `proximity.py`

Optimized iteration through `Document` objects as described in the demos in the `__init__.py` file of the `analysis` module

Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>

joshfeli committed:
Added a description to the `FieldError` `except` block

joshfeli committed:
Revert `app/analysis/__init__.py` to initial state

Also commented out unused/outdated `import` statements in `proximity.py`

joshfeli committed:
Merge branch 'main' into optimize_proximity_document_iteration

joshfeli committed:
Created a new `proximity.py` file

Refactoring from scratch!

joshfeli committed:
Implemented ideas for the core of `proximity.py`

The `run_analysis`, `generate_gender_token_counters`, and `generate_token_counter` functions

joshfeli committed:
Merge branch 'main' into optimize_proximity_document_iteration

joshfeli committed:
Removed all references to `tags` parameter

Instead of filtering the results by tag, we run the analysis using all `nltk` tags available and plan to let the user do the filtering on the frontend

Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>
Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>
Co-Authored-By: Ashar Farooq <ashar744@gmail.com>

joshfeli committed:
Deleted old `proximity.py` file

Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>
Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>
Co-Authored-By: Ashar Farooq <ashar744@gmail.com>

joshfeli committed:
Revert "Deleted old `proximity.py` file"

This reverts commit 8a54005d1051baee4b09619449ac2a42ed84128e.

joshfeli committed:
Revert "Revert "Deleted old `proximity.py` file""

This reverts commit 69396b03188d039f452dc48fb890d41190fa3919.

sharabhuiyan commented on backend/app/analysis/proximity.py:
```diff
@@ -0,0 +1,85 @@
+
+    for gender_id in gender_ids:
+        doc_result = generate_token_counter(text_query, gender_id, word_window)
+        results.update({Gender.objects.values_list('label', flat=True).get(pk=gender_id): doc_result})
```
Good use of lazy evaluation! I was wondering if it may be more efficient to add in dictionary entries using `dict[key] = value` (see below) because your results dictionary is empty and you are adding new entries as you iterate over the IDs. From my understanding, `dict.update()` looks to see if the key already exists in the dictionary and if it does it overwrites that key-value pair or else it just adds the dictionary entry. Let me know what you think!
```suggestion
       gender_label = Gender.objects.values_list('label', flat=True).get(pk=gender_id)
       results[gender_label] = doc_result
```

joshfeli replied:
Thanks for the suggestion! I changed that line as you suggested, but I also just hit the database to load the `Gender` object in `generate_gender_token_counters` instead of `generate_token_counters`. After talking with @MBJean, we concluded that each `Gender` model is slim enough to hit the database earlier rather than later and save ourselves an extra query.

sharabhuiyan reviewed (approved):
Good job! This PR looks good to me! I just left a comment about something that I was wondering, please let me know your thoughts.

joshfeli committed:
Reformatted `dict` update in `generate_gender_token_counters`

I also decided to pull the `Gender` object from the database earlier in the `generate_gender_token_counters` function rather than in `generate_token_counters`. The `Gender` objects are slim enough that we don't have to be so careful about lazy evaluation: in this case, this reduces the need for an extra query to hit the database just to get the `label` field value.

Co-Authored-By: sharabhuiyan <sharabhuiyan725@gmail.com>

joshfeli committed:
Merge branch 'main' into optimize_proximity_document_iteration

joshfeli committed:
Refactored `proximity.py` to accept the id of a `Corpus`

The `Corpus` id will probably come from a `views.py` call to `run_analysis`.

Co-Authored-By: sharabhuiyan <sharabhuiyan725@gmail.com>

joshfeli committed:
Removed debugging packages from `requirements.txt`

MBJean commented on backend/app/analysis/proximity.py:
```diff
@@ -0,0 +1,88 @@
+)
+
+
+def run_analysis(corpus_id, gender_ids, word_window):
```
I think we can simply pass the `Corpus` and `Gender` instances here. I can imagine that there'll be an API layer that performs some basic validation to ensure that the `Gender` and `Corpus` instances specified by the API input (say, a user passes down the appropriate `id` for each from the front end) actually do exist. That validation will necessarily require the look up for each of these instances, so passing them down here will save us using `Corpus.objects.filter` below (i.e., it can be come `corpus.filter(etc...)`. We can also pass in the `Gender` instances instead of their `id`s, turning `Gender.objects.get` into simply `gender` (vel sim.).

joshfeli replied:
@MBJean Our team was actually wondering about this. In the `transfer_proximity` branch, we created a new `ProximityAnalyses` model to store the results of these analyses in the database and used it as a sort of caching mechanism in `proximity.py` (we're about to make a PR). In that branch, we modified the three functions this PR presents, and one of the modifications was scrapping the `gender_ids` paramater. We figured it might be better to run the analysis for all `Gender` objects in the database and leave it to frontend to filter the display as per the user's preferences. I've changed the code so that we now pass in a set of all `Gender` objects instead of the list of `gender_ids`.

Regarding the `Corpus` object, we were unsure if loading the `Corpus` object would also load all associated `Document` objects from the database. Without the `Document`s, each `Corpus` object is pretty slim, so we could probably also implement that validation (possibly using the [`.exists()`](https://docs.djangoproject.com/en/3.1/ref/models/querysets/#exists) query) in a view function.

joshfeli replied:
Update about that new PR I mentioned: with these new changes, we'll likely have to go back and rethink our implementation for the other functions in `proximity.py` as well as how we will connect this to an API endpoint.

MBJean commented on backend/app/analysis/proximity.py:
```diff
@@ -0,0 +1,88 @@
+            Document.objects.values_list('tokenized_text', flat=True).filter(pk=key),
+            gender_ids,
+            word_window
+        )
```
Could you remind me what the value is in passing the `text_query` rather than the `tokenized_text` itself? That is: evaluate the `tokenized_text` here in line 28 or so, then pass that in. As long as we perform the lookup inside the loop that starts on line 27, we should achieve the same effect as the lazy-loading, right?

joshfeli replied:
@MBJean Ah, that makes much more sense! I had originally though of delaying loading the text until the last minute, but I see that this implementation keeps hitting the database for the same list of strings, all in just one iteration through the `doc_ids`. I've changed it so that we load the text at line 28 or so.

MBJean commented on backend/app/analysis/proximity.py:
```diff
@@ -0,0 +1,88 @@
+            tagged_tokens = nltk.pos_tag(words)
+            for tag_index, _ in enumerate(tagged_tokens):
+                word = words[tag_index]
+                output[word] += 1
```
I _think_ this is pretty close, but unless I'm misreading this, this doesn't differentiate what `pos_tag` each word is labeled as, right? I'm thinking that this output here might need to be a dictionary that keys part-of-speech tags to `Counter` instances. This would be a difference in from the original project. Something like:

```
output = {
  'NN': Counter({ 'door': 2, 'shovel': 1, etc.}),
  'JJ': Counter({'happy': 3, 'sad': 1, etc.})
}
```

kenalba replied:
A secondary question on this snippet of code -- since we've already done our `pos_tag` operation as part of document initialization (we have both `doc.tokenized_text` and `doc.part_of_speech_tags`, can we / should we pass those along and save ourselves the (relatively) expensive nltk.pos` operation?

irinazoccolini replied:
That makes sense. We've changed it so that we pass along `doc.part_of_speech_tags` instead of the tokenized text and don't have to use nltk.pos. We also changed ouput to a dictionary that keys part-of-speech tags to `Counter `instances.

MBJean reviewed (changes requested):
This is looking great so far! I have a few comments above that I'd interested in getting your feedback on.

joshfeli committed:
Altered lazy-loading of `Document`s; removed `gender_ids` as a parameter

Instead, the analysis will run for all `Gender` objects in the database, leaving further filtering to the frontend for displaying results.

joshfeli committed:
Merge branch 'main' into optimize_proximity_document_iteration

farooqashar committed:
Improved the `generate_token_counter` method

Revised the output to be a dictionary mapping a part of speech to a Counter instance instead of just a Counter instance. Removed the need to call pos_tag.

Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>
Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>
Co-Authored-By: Joshua Feliciano <41080599+joshfeli@users.noreply.github.com>

MBJean commented on backend/app/analysis/proximity.py:
```diff
@@ -0,0 +1,87 @@
+    :return: A dict mapping `Document` ids to a dict mapping strings (`Gender` labels) to a `Counter` instance.
+    """
+    results = {}
+    genders = set(Gender.objects.all())
```
Ultimately, we'll want the user to be able to specify which genders they want to see as part of this analysis. For now, though, I think grabbing `.all` is fine.

MBJean commented on backend/app/analysis/proximity.py:
```diff
@@ -0,0 +1,87 @@
+
+    for gender in genders:
+        doc_result = generate_token_counter(pos_tags, gender, word_window)
+        results[gender.label] = doc_result
```
We can probably use `results[gender]` here. Using the `.label` was for readability when the user was engaging with this data only in the terminal. Here we can leave the cleaning up (i.e., making it readable to the user) to the view layer.

MBJean reviewed (commented):
This is looking great. One last thought: what if we later want to know which words are in a window around, specifically, a `Gender`'s `subject` identifier? I _think_, to do that, we'd need to create another layer in our output dictionary: i.e., 
```
{
  <Gender>:
    <subject>: {
      <pos_tag>: Counter({...})
    },
    <object>: {
      <pos_tag>: Counter({...})
    },
    <other>: {
      <pos_tag>: Counter({...})
    }
}
```
vel sim.

Any thoughts on the above? It seems like, if we end up with that kind of data structure, any kind of transformation would be very straightforward (i.e., give me all of the words near a specific `Gender`'s subject in a given text, or multiple texts).

joshfeli committed:
Added a new layer to the `results` dictionary

This new layer maps `Gender` objects to a dictionary that maps _pronoun types_ to the nltk_tag:Counter mapping previously tied with the `Gender` instances. As @MBJean suggested, data transformations would be much more straightforward.

Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>
Co-Authored-By: Ashar Farooq <ashar744@gmail.com>
Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>

joshfeli committed:
Working out a `TestCase` for the functions in `proximity.py`

Co-Authored-By: Watermelanie <46464201+Watermelanie@users.noreply.github.com>
Co-Authored-By: Ashar Farooq <ashar744@gmail.com>
Co-Authored-By: irinazoccolini <71565864+irinazoccolini@users.noreply.github.com>

joshfeli committed:
Added a migration to modify the database seed

(As described in `migrations/0006_initial_data.py`)

joshfeli committed:
Added an extra space

joshfeli committed:
Focused on revising expected results for text 1

joshfeli committed:
Merge branch 'main' into optimize_proximity_document_iteration

joshfeli committed:
Merge branch 'modify_db_seed' into optimize_proximity_document_iteration

joshfeli committed:
Added tagged tokens for the three texts in comments

Also made the `PronounSeries` and `Gender` classes hashable after getting an error when putting a `Gender` object in a set

joshfeli committed:
Removed a pronoun from its own word window

(To avoid counting a pronoun in its own `Counter`)

joshfeli committed:
Added padding to window

Ensures that the first and last `word_window` tokens in `pos_tags` are considered (they would not be without the padding).

MBJean commented on backend/app/tests.py:
```diff
@@ -208,3 +212,163 @@ def test_add_document_to_corpus(self):
+
+    def test_proximity(self):
+        male = Gender.objects.get(pk=1)
+        female = Gender.objects.get(pk=2)
```
We may want to reference these by `label` rather than primary key, as there's a hypothetical where the primary key isn't consistent based on the seeds. I mean, that's true of the label as well, but it feels safer.

joshfeli replied:
Good point! I initially avoided referencing by `label` in the case that someone created similar constants and saved them to their local database (at least I have, so the label `Male` would return more than one object). I've referenced by `label` in addition to `pk` just to be safe. Since the newly created database for each `TestCase` should be empty except for the data in the migration files, the `.get()` lookup would only return a `Gender.DoesNotExist` error if we change migrations `0006` and `0009` (`0009` is in PR #48, but I've checked to make sure that it doesn't affect the `pk` ordering of any objects).

joshfeli committed:
Focused on debugging tests for `text_3`

joshfeli committed:
Focused on debugging tests for `text_2`

joshfeli committed:
Identified a possible issue

(Marked in lines 442-446)

MBJean reviewed (approved):
This is looking fantastic! Great work, all. You've created a straightforward and performant pattern for parsing a `Document` that we can use in most of our analysis modules going forward, and I have no further comments. I can see that you've got one unresolved testing issue, but I'll leave my approval here for when you do resolve it. Great work!

joshfeli committed:
Merge branch 'main' into optimize_proximity_document_iteration

joshfeli committed:
Resolved identified issue

It came down to an `nltk` discrepancy in tagging one `'most'` token as an adjective vs an adverb!
